File: collatz.c
#include <stdio.h>
#include <math.h>


///////////#########################
void main() {
  long long n_curr, n_min, sum_of_mins, n_min_init=999999, n_max = 1000000000, div=4;
  int a, b, step, escaped, n_curr_iter, nlim=20;
  /*All quadrants*/
  int amin=-1024 , amax=1024, bmin=-1024 , bmax=1024;
  int counter_escaped, counter_ints;
  double min_fix[nlim], escape_value = 3.14159;
  int calc_steps=10, max_mini_steps = 100;


  /*Collatz mapper to print to external file*/
  for (a = amin; a < amax; a++) {
    for (b = bmin; b < bmax; b++) {
      /**/
      for (n_curr_iter = 1; n_curr_iter < nlim+1; n_curr_iter++) {
        //printf("a, b, n = %d, %d, %lld\n", a, b, n_curr_iter);
        n_curr = n_curr_iter;
        escaped = 0;
        /*Calculating stage*/
        for (size_t i = 0; i < calc_steps; i++) {
          if ( fmod(n_curr,div)==0 ){
            if (n_curr == 0) {
              n_min = 0;
              break;
            } else {
              n_curr = n_curr / div;
            }
          } else {
            n_curr = a*n_curr + b;
          }
        }
        step = 1;
        n_min = n_curr + n_min_init;
        /*Minimization stage*/
        while (n_min != n_curr){
          //printf("n_curr = %lld, n_min = %lld\n", n_curr, n_min);
          if (n_min > n_curr) {
            n_min = n_curr;
          }
          if ( fmod(n_curr,div)==0 ) {
            if (n_curr == 0) {
              n_min = 0;
              break;
            } else {
              n_curr = n_curr / div;
            }
          } else {
            n_curr = a*n_curr + b;
          }
          step++;
          if (n_curr < -n_max || n_curr > n_max || step > max_mini_steps) {
            //printf("escaped\n");
            escaped = 1;
            break;
          }
        }
        /*Adding to the array of smallest fixed points*/
        if (escaped) {
          min_fix[n_curr_iter-1] = escape_value;
          escaped = 0;
        } else {
          min_fix[n_curr_iter-1] = 1.0*n_min;
        }
      }
      /*Finding convergence and adding point to map*/
      counter_escaped=0;
      counter_ints=0;
      sum_of_mins=0;
      for (size_t i = 0; i < nlim; i++) {
        if (min_fix[i] == escape_value) {
          counter_escaped++;
        }
        else {
          sum_of_mins += min_fix[i];
          counter_ints++;
        }
      }
      if (counter_escaped>0) {
        if (counter_escaped==nlim) {
          printf("%d\t", 0);
        } else {
          printf("%d\t", 1);
        }
      } else {
        if (1.0*sum_of_mins/counter_ints == min_fix[0]) {
          printf("%d\t", 3);
        } else {
          printf("%d\t", 2);
        }
      }
    }
    printf("\n");
  }
}

File: collatz.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import matplotlib.animation as anim
import sys

def collatz_mapper(amax, bmax, div=2, nlim=20):
    """
    Returns a Numpy array with the Collatz Map. Utilizes ints to build the map.

    Parameters
    ----------
    amax: int
        Maximum positive 'a' value of the map (horizontal axis). The map goes
        from -amax to amax.
    bmax: int
        Maximum positive 'b' value of the map (vertical axis). The map goes
        from -bmax to bmax.
    div: int
        Divisor for the generalized Collatz Problem.
    nlim: int
        Maximimum number of integers to search for convergence.
    """
    a_range = range(-amax, amax)
    b_range = range(-bmax,bmax)
    d = div

    collatz_map = np.zeros((len(b_range), len(a_range)))

    n_currs = range(1,nlim+1)
    n_min_init = 999999 #initialization value for n_min, so that it is bigger than n_curr
    n_max = 1000000000
    calc_steps = 10
    max_mini_steps = 100 #maximum steps in collatz sequence to check for convergence
    escape_value = np.pi #value given to list of fixed points if number of steps exceded.

    for a in a_range:
        for b in b_range:
            min_fix = []
            for n_curr in n_currs:
                #print('a, b, n = ',a, b, n_curr)
                step = 1
                escaped = False
                for _ in range(calc_steps):
                    if n_curr%d == 0:
                        if n_curr == 0:
                           n_min = 0
                           break
                        else:
                           n_curr = n_curr // d
                    else:
                        n_curr = a*n_curr + b
                n_min = n_curr + n_min_init
                while n_min != n_curr:
                    #print('n_curr=', n_curr, ', n_min=', n_min)
                    n_min = min(n_min, n_curr)
                    if n_curr%d == 0:
                        if n_curr == 0:
                           n_min = 0
                           break
                        else:
                           n_curr = n_curr // d
                    else:
                        n_curr = a*n_curr + b
                    step += 1
                    if step > max_mini_steps or abs(n_curr) > n_max:
                        #print('escaped')
                        escaped = True
                        break
                if escaped:
                    min_fix.append(escape_value)
                    escaped = False
                else:
                    min_fix.append(n_min)
            #adding point to map
            min_fix_setlist = list(set(min_fix))
            if np.pi in min_fix_setlist:
                if [np.pi]==min_fix_setlist:
                    collatz_map[b + bmax, a + amax] = 0
                else:
                    collatz_map[b + bmax, a + amax] = 1
            else:
                if len(min_fix_setlist) == 1:
                    collatz_map[b + bmax, a + amax] = 3
                else:
                    collatz_map[b + bmax, a + amax] = 2
    #
    return collatz_map
#
def collatz_plotter(col_map, div = 0, cmap='inferno', legend=True):
    """
    Plots the collatz map.

    Parameters
    ----------
    colmap: numpy array
        Contains the map.
    div: int
        Used as figure number and for filename if empty.
    save_fig: boolean
        Declares whether to save the figure or not.
    savename: string
        Name of the file where figure will be saved.
    cmap: string or colormap
        Set by default to 'inferno'.
    """
    if div==0:
        fig = plt.figure()
        div = ''
    else:
        fig = plt.figure(div)
    ax=fig.add_subplot(1,1,1)
    values = np.unique(col_map)
    image = plt.imshow(col_map, cmap=cmap)
    colors = [image.cmap(image.norm(value)) for value in values]
    labels = ['no convergence', 'convergence for some', 'convergence for all at diff', 'convergence for all at same']
    patches = [mpatches.Patch(color=colors[i], label=labels[i]) for i in range(len(values))]
    if legend:
        plt.legend(handles=patches, bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)
    if div!=0:
        plt.title('divisor = '+str(div))
    plt.axis('off')
    plt.show()
#
def collatz_animation(amax, bmax, div):
    """
    Returns an animation object and saves it as Gif.
    """
    pass
#
if __name__ == '__main__':
    int_parameters = []
    for arg in sys.argv[1:]:
        int_parameters.append(int(arg))
    if all([type(elem)==int for elem in int_parameters]):
        amax, bmax, div, nlim = int_parameters[:4]
    else:
        amax, bmax, div, nlim = eval(input('enter all ints with commas: amax, bmax, div, nlim\n'))
    #amax, bmax, div, nlim = 1024, 1024, 2, 20
    print('parameters: amax=', amax, ', bmax=',bmax, ', div=',div, ', nlim=',nlim)
    collatz_map = collatz_mapper(amax, bmax, div, nlim)
    #Save collatz map to file
    filename = 'collatz_map_a'+str(amax)+'_b'+str(bmax)+'_div'+str(div)+'.colmap'
    with open(filename, 'w') as file:
        for row in range(2*amax):
            for column in range(2*bmax):
                file.write(str(collatz_map[row,column])+'\t')
            file.write('\n')
    #
    collatz_plotter(collatz_map, div)
